velox/exec/GroupingSet.cpp

```diff
@@ -598,6 +606,73 @@ bool GroupingSet::getGlobalAggregationOutput(
   return true;
 }
 
+bool GroupingSet::getDefaultGlobalGroupingSetOutput(
+    int32_t batchSize,
+    bool isPartial,
+    RowContainerIterator& iterator,
+    RowVectorPtr& result) {
+  VELOX_CHECK_LE(globalGroupingSets_->size(), batchSize);
+
+  // Global aggregates don't have grouping keys. But global grouping sets
+  // have null values in grouping keys and a groupId column as well. These
+  // key fields precede the aggregate columns in the result.
+  // This logic builds a row with just aggregate fields to reuse the global
+  // aggregate computation from the regular GroupingSet code-path.
+  auto outputType = asRowType(result->type());
+  auto firstAggregateCol = outputType->children().size() - aggregates_.size();
+  std::vector<std::string> names;
+  std::vector<TypePtr> types;
+  names.reserve(aggregates_.size());
+  types.reserve(aggregates_.size());
+  for (int i = firstAggregateCol; i < outputType->children().size(); i++) {
+    names.push_back(fmt::format("agg{}", i));
+    types.push_back(outputType->childAt(i));
+  }
+  auto aggregatesType =
+      std::make_shared<RowType>(std::move(names), std::move(types));
+  auto globalAggregatesRow = std::dynamic_pointer_cast<RowVector>(
+      BaseVector::create(aggregatesType, 1, &pool_));
+
+  auto status =
+      getGlobalAggregationOutput(1, isPartial, iterator, globalAggregatesRow);
+  if (!status) {
+    return false;
+  }
+
+  // There is one output row for each global GroupingSet.
+  result->resize(globalGroupingSets_->size());
+  // These first columns are for grouping keys (which could include the
+  // GroupId column). For a global grouping set row :
+  // i) Non-groupId grouping keys are null.
+  // ii) GroupId column is populated with the global grouping set number.
+  for (auto i = 0; i < firstAggregateCol; i++) {
+    if (groupIdChannel_.has_value() && i == groupIdChannel_.value()) {
```

Comment:
Is it possible that groupIdChannel_.has_value() is false?

Reply:
Its possible that groupIdChannel is never false. But I was trying to be conservative about it here.

Reply:
"defensive coding" is an anti pattern; let's not do that